### PR TITLE
[XLA:GPU] add workspace rewrite for FP8 gemm

### PR DESCRIPTION
**Pull request by shawnwang18**
Add  workspace rewrite for FP8 with amax fusion.

### CODE REVIEW

File: xla/service/gpu/gemm_rewriter.cc [hunk 1]
```diff
@@ -2314,7 +2325,7 @@ class GemmWorkspaceRewriteVisitor : public DfsHloRewriteVisitor {
       custom_call->set_output_to_operand_aliasing({{{0}, {2, {}}}});
     }
 
-    if (has_aux_output) {
+    if (instr->shape().IsTuple()) {
       for (auto user : instr->users()) {
```
Comment (reedwm) [inline xla/service/gpu/gemm_rewriter.cc:2329]: Can you replace this entire for-loop with `ReplaceInstr(instr, new_call)`? That should cause all the GetTupleElement instructions to refer to new_call insteade of instr.

I just noticed this but it's unrelated to the fix. So if this is replaced by `ReplaceInstr(instr, new_call)`, it should be done in a separate PR

File: xla/service/gpu/ir_emitter_unnested.cc [hunk 2]
```diff
@@ -55,17 +55,17 @@ limitations under the License.
 #include "llvm/IR/LLVMContext.h"
 #include "llvm/IR/Module.h"
 #include "llvm/Linker/Linker.h"
-#include "mlir/AsmParser/AsmParser.h"  // from @llvm-project
-#include "mlir/Dialect/LLVMIR/LLVMDialect.h"  // from @llvm-project
+#include "mlir/AsmParser/AsmParser.h"               // from @llvm-project
```
Comment (reedwm) [inline xla/service/gpu/ir_emitter_unnested.cc:58]: Why have this unnecessary formatting? Did an autoformatter do this?

**Review by reedwm — CHANGES_REQUESTED** [review]

File: xla/service/gpu/backend_configs.proto [hunk 3]
```diff
@@ -104,6 +104,7 @@ message GemmBackendConfig {
 
   optional bool grad_x = 16;
   optional bool grad_y = 17;
+  optional bool has_damax_output = 18;
```
Comment (philipphack) [inline xla/service/gpu/backend_configs.proto:107]: Why is this an optional bool? Do we need three different states?

File: xla/service/gpu/gemm_rewriter.cc [hunk 4]
```diff
@@ -2261,10 +2268,14 @@ class GemmWorkspaceRewriteVisitor : public DfsHloRewriteVisitor {
       TF_ASSIGN_OR_RETURN(
           has_aux_output,
           xla::gpu::gpublas_lt::EpilogueHasAuxiliaryOutput(epilogue));
+      if (config.has_has_damax_output()) {
```
Comment (philipphack) [inline xla/service/gpu/gemm_rewriter.cc:2271]: See above. I suggest making this a non-optional bool or renaming it to damax_output.

File: xla/service/gpu/ir_emitter_unnested.cc [hunk 5]
```diff
@@ -758,9 +758,8 @@ absl::Status IrEmitterUnnested::EmitCublasLtMatmulThunkF8(
 
   TF_ASSIGN_OR_RETURN(bool has_vector_bias,
                       xla::gpu::gpublas_lt::EpilogueAddsVectorBias(epilogue));
-  bool has_damax = instr->shape().IsTuple();
   xla::ShapeIndex output_index =
-      has_damax ? xla::ShapeIndex{0} : xla::ShapeIndex{};
+      instr->shape().IsTuple() ? xla::ShapeIndex{0} : xla::ShapeIndex{};
```
Comment (philipphack) [inline xla/service/gpu/ir_emitter_unnested.cc:762]: Are there situations where an FP8 custom call does not return a tuple?

File: xla/service/gpu/ir_emitter_unnested.cc [hunk 6]
```diff
@@ -823,9 +823,7 @@ absl::Status IrEmitterUnnested::EmitCublasLtMatmulThunkF8(
   TF_RET_CHECK(!has_aux_output);
   std::optional<BufferAllocation::Slice> workspace_buffer;
   if (instr->shape().IsTuple() &&
-      (instr->shape().tuple_shapes_size() - has_aux_output - 1)) {
-    TF_RET_CHECK((has_aux_output && instr->shape().tuple_shapes_size() == 3) ||
-                 (!has_aux_output && instr->shape().tuple_shapes_size() == 2));
+      (instr->shape().tuple_shapes_size() - has_damax == 2)) {
```
Comment (philipphack) [inline xla/service/gpu/ir_emitter_unnested.cc:826]: IIUC, the call always returns a workspace. Do we need these conditionals?

**Review by philipphack — CHANGES_REQUESTED** [review]

File: xla/service/gpu/backend_configs.proto (hunk 3 above)
Comment (shawnwang18) [reply xla/service/gpu/backend_configs.proto:107]: Without optional, it will require protobuf to initialize the has_damax_output to false. though protobuf does it, but I think have an optional state should make the code more clear.

File: xla/service/gpu/gemm_rewriter.cc (hunk 1 above)
Comment (shawnwang18) [reply xla/service/gpu/gemm_rewriter.cc:2329]: Yeah, let's do it in a separate PR.

File: xla/service/gpu/ir_emitter_unnested.cc (hunk 5 above)
Comment (shawnwang18) [reply xla/service/gpu/ir_emitter_unnested.cc]: I fixed the code that assume the output shape is tuple.

File: xla/service/gpu/gemm_rewriter.cc [hunk 7]
```diff
@@ -2261,10 +2268,14 @@ class GemmWorkspaceRewriteVisitor : public DfsHloRewriteVisitor {
       TF_ASSIGN_OR_RETURN(
           has_aux_output,
           xla::gpu::gpublas_lt::EpilogueHasAuxiliaryOutput(epilogue));
+      if (config.has_damax_output()) {
+        has_damax_output = config.damax_output();
```
Comment (reedwm) [inline xla/service/gpu/gemm_rewriter.cc:2272]: No need to check if the config has damax_output, you can just set `has_damax_output` uncondtionally. 

I don't think you even need a `has_damax_output` local variable. You can just use `config.damax_output()` in the if-statement below

**Review by reedwm — CHANGES_REQUESTED** [review]

File: xla/service/gpu/ir_emitter_unnested.cc [hunk 8]
```diff
@@ -822,10 +822,7 @@ absl::Status IrEmitterUnnested::EmitCublasLtMatmulThunkF8(
   BufferAllocation::Slice aux;  // Not used.
   TF_RET_CHECK(!has_aux_output);
   std::optional<BufferAllocation::Slice> workspace_buffer;
-  if (instr->shape().IsTuple() &&
-      (instr->shape().tuple_shapes_size() - has_aux_output - 1)) {
-    TF_RET_CHECK((has_aux_output && instr->shape().tuple_shapes_size() == 3) ||
-                 (!has_aux_output && instr->shape().tuple_shapes_size() == 2));
+  if (instr->shape().tuple_shapes_size() - config.damax_output() == 2) {
```
Comment (philipphack) [inline xla/service/gpu/ir_emitter_unnested.cc:825]: Is there a known case where this is false? If not, this is probably not needed.

Comment (shawnwang18) [reply xla/service/gpu/ir_emitter_unnested.cc:825]: I don't know a case, but the workspace rewrite has a guard like this: 

      if (!((instr->shape().IsTuple() &&
             instr->shape().tuple_shapes_size() ==
                 has_aux_output + config.damax_output() + 1) ||
            instr->shape().IsArray())) {
       return absl::OkStatus();
}

If gemm_rewrite happens to create a new case where tuple includes other outputs, then it is an unsupported case.

**Review by reedwm — APPROVED** [review]
